src/cargo/sources/registry/remote.rs

```diff
@@ -187,84 +186,146 @@ impl<'cfg> RegistryData for RemoteRegistry<'cfg> {
         Ok(())
     }
 
-    fn download(&mut self, pkg: &PackageId, checksum: &str)
-                -> CargoResult<FileLock> {
-        let filename = format!("{}-{}.crate", pkg.name(), pkg.version());
-        let path = Path::new(&filename);
+    fn download(&mut self, pkgs_with_checksum: &[(&PackageId, &str)])
+                -> CargoResult<Vec<FileLock>> {
+        let mut result = Vec::new();
+        let mut to_download = Vec::new();
+        for (idx, &(pkg, _)) in pkgs_with_checksum.iter().enumerate() {
+            let filename = format!("{}-{}.crate", pkg.name(), pkg.version());
+            let path = Path::new(&filename);
 
-        // Attempt to open an read-only copy first to avoid an exclusive write
-        // lock and also work with read-only filesystems. Note that we check the
-        // length of the file like below to handle interrupted downloads.
-        //
-        // If this fails then we fall through to the exclusive path where we may
-        // have to redownload the file.
-        if let Ok(dst) = self.cache_path.open_ro(path, self.config, &filename) {
+            // Attempt to open an read-only copy first to avoid an exclusive write
+            // lock and also work with read-only filesystems. Note that we check the
+            // length of the file like below to handle interrupted downloads.
+            //
+            // If this fails then we fall through to the exclusive path where we may
+            // have to redownload the file.
+            if let Ok(dst) = self.cache_path.open_ro(path, self.config, &filename) {
+                let meta = dst.file().metadata()?;
+                if meta.len() > 0 {
+                    result.push(dst);
+                    continue;
+                }
+            }
+            let mut dst = self.cache_path.open_rw(path, self.config, &filename)?;
```

Comment:
It is possible to deadlock in our case, as we're locking multiple files at once. The solution with least overhead is to have a consistent order on package queue, e.g. using `BTreeMap` for `PackageSet`.

(Given that `PackageSet` may also be refactored out, I'm not sure what the final solution would be though.)

Reply:
Ah good point! With parallel downloads though there may not be much point in locking per-package, I'd be fine with changing the registry to acquire a global lock when writing (which is append only) and then each crate file only exists if it's actually valid